File: libs/power-log-parser/src/lib/models/game.ts
import { Action } from './action';
import { GameData } from './game-data';

export class Game {
	TimeStamp: string = '';
	BuildNumber: number = 0;
	Type: number = 0;
	GameType: number = -1;
	FormatType: number = -1;
	ScenarioID: number = 0;
	GameSeed: number = 0;
	Data: GameData[] = [];

	AddData(data: GameData): void {
		this.Data.push(data);
	}

	FilterGameData(...types: (new (...args: any[]) => GameData)[]): GameData[] {
		const result: GameData[] = [];
		for (const data of this.Data) {
			result.push(data);
			this.ExtractData(result, data);
		}
		if (types.length === 0) {
			return result;
		}
		return result.filter((data) => types.some((t) => data instanceof t));
	}

	ExtractData(result: GameData[], data: GameData): void {
		if (data instanceof Action) {
			for (const gameData of data.Data) {
				result.push(gameData);
				if (data !== gameData) {
					gameData.InternalParent = data;
				}
				this.ExtractData(result, gameData);
			}
		}
	}

	GetLastAction(predicate: (action: Action) => boolean): Action | null {
		return this.GetLastActionInternal(this.Data, predicate);
	}

	private GetLastActionInternal(data: GameData[], predicate: (action: Action) => boolean): Action | null {
		for (let i = data.length - 1; i >= 0; i--) {
			if (!(data[i] instanceof Action)) {
				continue;
			}
			const action = data[i] as Action;
			const childMatch = this.GetLastActionInternal(action.Data, predicate);
			if (childMatch != null) {
				return childMatch;
			}
			if (predicate(action)) {
				return action;
			}
		}
		return null;
	}
}

export class HearthstoneReplay {
	Build: string = '';
	Version: string = '';
	Games: Game[] = [];
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/ReplayData/Game.cs
#region

using System.Linq;
using System.Collections.Generic;
using System.Xml.Serialization;
using HearthstoneReplays.Parser.ReplayData.Entities;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using HearthstoneReplays.Parser.ReplayData.Meta;
using HearthstoneReplays.Parser.ReplayData.Meta.Options;
using System;

#endregion

namespace HearthstoneReplays.Parser.ReplayData
{
	public class Game
	{
        public readonly object listLock = new object();

        [XmlIgnore]
		public DateTime TimeStamp { get; set; }

        [XmlAttribute("ts")]
        public string TsForXml
        {
            get
            {
                var hours = TimeStamp.Hour;
                if (hours < ReplayParser.start.Hour)
                {
                    hours += 24;
                }
                var timestampString = this.TimeStamp.ToString("HH:mm:ss.ffffff");
                var split = timestampString.Split(':');
                return ("" + hours).PadLeft(2, '0') + ":" + split[1] + ":" + split[2];
            }
            set => this.TimeStamp = DateTime.Parse(value);
        }

        [XmlAttribute("buildNumber")]
		public int BuildNumber { get; set; }

		[XmlAttribute("type")]
		public int Type { get; set; }

		[XmlAttribute("gameType")]
		public int GameType { get; set; }

		[XmlAttribute("formatType")]
		public int FormatType { get; set; }

		[XmlAttribute("scenarioID")]
		public int ScenarioID { get; set; }

		[XmlAttribute("gameSeed")]
		public int GameSeed { get; set; }

		[XmlElement("Block", typeof(GameActions.Action))]
		[XmlElement("Choices", typeof(Choices))]
		[XmlElement("FullEntity", typeof(FullEntity))]
		[XmlElement("GameEntity", typeof(GameEntity))]
		[XmlElement("ChangeEntity", typeof(ChangeEntity))]
		[XmlElement("ShowEntity", typeof(ShowEntity))]
		[XmlElement("HideEntity", typeof(HideEntity))]
		[XmlElement("Options", typeof(Options))]
		[XmlElement("Player", typeof(PlayerEntity))]
		[XmlElement("SendChoices", typeof(SendChoices))]
		[XmlElement("SendOption", typeof(SendOption))]
		[XmlElement("TagChange", typeof(TagChange))]
		[XmlElement("MetaData", typeof(MetaData))]
		[XmlElement("ChosenEntities", typeof(ChosenEntities))]
		[XmlElement("ShuffleDeck", typeof(ShuffleDeck))]
		public List<GameData> Data { get; set; }

        public Game()
		{
			Data = new List<GameData>();
            FormatType = -1;
            GameType = -1;
        }

        public void AddData(GameData data)
        {
            lock (listLock)
            {
                Data.Add(data);
            }
        }

		/// <summary>
		/// CAUTION: it builds the whole data and return it to you.
		/// So if you're interested in getting a specific node, it's probably best to iterate over it yourself
		/// </summary>
		/// <param name="types"></param>
		/// <returns></returns>
		internal List<GameData> FilterGameData(params System.Type[] types)
		{
			// Build the list - could probably be built incrementally instead of rebuilding it completely every time
			List<GameData> result = new List<GameData>();
            lock (listLock)
            {
			    foreach (GameData data in Data)
			    {
				    result.Add(data);
				    ExtractData(result, data);
			    }

			    // Now filter it
			    return result.Where(data => types == null || types.Contains(data.GetType())).ToList();
            }
		}

        internal void ExtractData(List<GameData> result, GameData data)
		{
			if (data.GetType() == typeof(GameActions.Action))
			{
				foreach (GameData gameData in ((GameActions.Action) data).Data)
				{
					result.Add(gameData);
					if (data != gameData)
					{
						gameData.InternalParent = data;
					}
					ExtractData(result, gameData);
				}
			}
		}

		public GameActions.Action GetLastAction(Predicate<GameActions.Action> predicate)
		{
			lock (listLock)
			{
				return GetLastActionInternal(Data, predicate);
			}
		}

        private GameActions.Action GetLastActionInternal(List<GameData> GameData, Predicate<GameActions.Action> predicate)
        {
            lock (listLock)
            {
                for (var i = GameData.Count - 1; i >= 0; i--)
                {
                    if (GameData[i].GetType() != typeof(GameActions.Action))
                    {
                        continue;
                    }
                    var action = GameData[i] as GameActions.Action;
					var childMatch = GetLastActionInternal(action.Data, predicate);
                    if (childMatch != null)
                    {
						return childMatch;
                    }
                    if (predicate(action))
                    {
                        return action;
                    }
                }
				return null;
            }
        }
    }
}